src/django_otp/models.py

```diff
@@ -293,6 +306,68 @@ class VerifyNotAllowed:
     N_FAILED_ATTEMPTS = 'N_FAILED_ATTEMPTS'
 
 
+class GenerationCooldownMixin(models.Model):
+    """
+    Mixin class for models requiring a cooldown duration between generating a challenge.
+    """
+
+    last_generated_timestamp = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text="The last time a token was generated for this device.",
+    )
+
+    def generate_is_allowed(self):
+        """
+        Checks if the time since the last token generation is greater than
+        configured (in seconds).
+        """
+        dt_now = timezone.now()
+        if (
+            not self.last_generated_timestamp
+            or (dt_now - self.last_generated_timestamp).total_seconds()
+            > self.get_cooldown_duration()
+        ):
+            return super().generate_is_allowed()
+        else:
+            return False, {
+                'reason': "COOLDOWN_DURATION_PENDING",
+                'next_generation_at': dt_now
```

Comment:
Should `dt_now` be `self.last_generated_timestamp` here?

Reply:
Indeed! I have corrected it and added a test for the message as well.